torch/_higher_order_ops/prim_hop_base.py

```diff
@@ -0,0 +1,178 @@
+        return PrimHOPBaseFunction.apply(self, subgraph, kwargs, *operands)
+
+    def _call_CompositeExplicitAutograd(self, subgraph, operands, *_, **kwargs):
+        from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+        mode = _get_current_dispatch_mode()
+        assert mode is None, "Mode should never be enabled for CPU/CUDA key"
+        return subgraph(*operands)
+
+    def _call_ProxyTorchDispatchMode(
+        self, proxy_mode, subgraph, operands, *_, **kwargs
+    ):
+        traced_graph = reenter_make_fx(subgraph)(*operands)
+        assert isinstance(proxy_mode.tracer, torch.fx.Tracer)
+        qualname = proxy_mode.tracer.get_fresh_qualname("subgraph")
+        proxy_mode.tracer.root.register_module(qualname, traced_graph)
+
+        node_args = (traced_graph, operands)
+        proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)  # type: ignore[attr-defined]
+        proxy_kwargs = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, kwargs)  # type: ignore[attr-defined]
+        out_proxy = proxy_mode.tracer.create_proxy(
+            "call_function", self, proxy_args, proxy_kwargs
+        )
+
+        out = self(subgraph, operands, **kwargs)
+        return track_tensor_tree(
+            out, out_proxy, constant=None, tracer=proxy_mode.tracer  # type: ignore[arg-type]
+        )
+
+    def _call_FakeTensorMode(self, mode, subgraph, operands, *_, **kwargs):
+        # TODO: this should probably route through FakeTensorMode to reuse caching
+        with mode:
+            return subgraph(*operands)
+
+    def _call_Functionalize(self, ctx, subgraph, operands, *_, **kwargs):
+        unwrapped_operands = ctx.unwrap_tensors(operands)
+        with ctx.redispatch_to_next() as m:
+            # We assume the subgraph doesn't mutate inputs and there is no aliasing.
+            # In the PT2 stack, this is Dynamo's responsibility to figure out.
+            functionalized_subgraph = FunctionWithNoFreeVars(
+                ctx.functionalize(subgraph)
+            )
+            out = self(functionalized_subgraph, unwrapped_operands, **kwargs)
+        return ctx.wrap_tensors(out)
+
+
+class PrimHOPBaseFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, hop, subgraph, kwargs, *operands):
+        ctx.hop = hop
+        ctx.operands = operands
+        ctx.subgraph = subgraph
+        ctx.kwargs = kwargs
+
+        with torch._C._AutoDispatchBelowAutograd():
+            return hop(subgraph, operands, **kwargs)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        subgraph = ctx.subgraph
+        operands = ctx.operands
+        kwargs = ctx.kwargs
+
+        # TODO: Something special needs to happen with min cut partitioner
+        with suspend_functionalization(), disable_functional_mode(), torch.enable_grad():
+            with disable_proxy_modes_tracing():
+                from .utils import _from_fun, create_fw_bw_graph
```

Comment:
I would like this to be invoke_subgraph `create_fw_bw_graph` (or replace utils.py one with the invoke_subgraph one). The main reason is the signatures for the joint graph and ergonomics that come with signatures.

For the joint graph from invoke_subgraph
1) Input signature is -  (*primals, *tangents).
2) Output signature is - (*grads, *fw_outs)

The nice thing about input signature is that it is already in a form that min-cut partitioner expects.

The nice thing about output signature is that `grads` are already at the right place. Once we apply min-cut partitioner, the new output signature of the backward graph would be just (*grads). So, the graph transformation is easy. All the getitems on the output of the old joint_graph are still valid for the new_bw graph. This reduces LOC and also chances of error.

Reply:
Yeah I can swap this out

Reply:
Updated to use invoke_subgraph's create_fw_bw_graph